File: Search-a-2D-Matrix/Solution.cpp
#include <vector>
using namespace std;

class Solution {
public:
    bool searchMatrix(vector<vector<int> > &matrix, int target) {
        if (matrix.size() == 0) {
            return false;
        }
        int n = matrix[0].size();

        for (int x, y, i = 0, j = matrix.size() * n, mid; i < j; ) {
            mid = (i + j) / 2;
            x = mid / n;
            y = mid % n;

            if (matrix[x][y] == target) {
                return true;
            }
            if (target < matrix[x][y]) {
                j = mid;
            } else {
                i = mid + 1;
            }
        }
        return false;
    }
};

int main() {
  return 0;
}

File: Search-a-2D-Matrix/Solution.py
class Solution:
    # @param matrix, a list of lists of integers
    # @param target, an integer
    # @return a boolean
    def searchMatrix(self, matrix, target):
        if len(matrix) == 0:
            return False
        n = len(matrix[0])
        i = 0
        j = n * len(matrix)
        while i < j:
            mid = (i + j) / 2
            x = mid / n
            y = mid % n
            if matrix[x][y] == target:
                return True
            if matrix[x][y] > target:
                j = mid
            else:
                i = mid + 1
        return False
